core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -211,32 +231,45 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
         rawConnectivityState = IDLE;
         updateBalancingState(IDLE, new RequestConnectionPicker(this));
         break;
+
       case CONNECTING:
         rawConnectivityState = CONNECTING;
         updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
         break;
+
       case READY:
         shutdownRemaining(subchannelData);
         addressIndex.seekTo(getAddress(subchannel));
         rawConnectivityState = READY;
         updateHealthCheckedState(subchannelData);
         break;
+
       case TRANSIENT_FAILURE:
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
-
-          // If no addresses remaining, go into TRANSIENT_FAILURE
-          if (!addressIndex.isValid()) {
-            helper.refreshNameResolution();
-            rawConnectivityState = TRANSIENT_FAILURE;
-            updateBalancingState(TRANSIENT_FAILURE,
-                new Picker(PickResult.withError(stateInfo.getStatus())));
+          if (addressIndex.increment()) {
+            cancelScheduleTask();
+            requestConnection(); // is recursive so might hit the end of the addresses
           }
         }
+
+        // If at end and: not doing happy eyeballs, just got to the end, or we have had enough TF
+        // since we got to the end, then refresh name resolution
+        if (!addressIndex.isValid()
+            && (!enableHappyEyeballs || numTf.get() >= addressIndex.size() || numTf.get() == 0)) {
```

Comment:
I see that you combined a few cases here. It's not trivial to verify and hard to maintain. 
For example, it looks wrong in the case of `enableHappyEyeballs==false`, because name resolution is refreshed every time when addressIndex is invalid and we have TF. 
Also I think increment the counter and the action should be together instead of scatter around. What about move L198-L200 down, and what about call `helper.refreshNameresolution` twice, once for the first time it goes invalid, then every time the counter > address size.